# Project: Wrangling and Analyze Data

## Data Gathering
In the cell below, gather **all** three pieces of data for this project and load them in the notebook. **Note:** the methods required to gather each data are different.
1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)

In [337]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns
import os
import requests
import tweepy
from io import StringIO
import json
from tqdm import tqdm


In [338]:
tweet_archive = pd.read_csv('twitter-archive-enhanced.csv') # read in the data

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [339]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
data = response.text
image_pred = pd.read_csv(StringIO(data), sep='\t')
image_pred.to_csv('image_predictions.tsv')

3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

In [340]:
from dotenv import load_dotenv
load_dotenv()

bearer_token = os.environ.get('BEARER_TOKEN')

tweet_id = list(tweet_archive['tweet_id'])
missing_tweets = []

In [341]:
# if not os.path.exists('tweet_json.txt'):
#     with open('tweet_json.txt', 'w'): pass
# def get_tweet():
#     auth = tweepy.OAuth2BearerHandler(bearer_token)
#     api = tweepy.API(auth)
#     for id in tqdm(tweet_id):
#         try:
#             tweet = api.get_status(id, tweet_mode='extended')
#             with open('tweet_json.txt', 'a') as f:
#                 json.dump(tweet._json, f)
#                 f.write('\n')
#         except:
#             print('Missing Tweet for id: {}'.format(id))
#             missing_tweets.append(id)
#             continue

# # Driver code
# if __name__ == '__main__':
# #   Call the function
#     get_tweet()


In [342]:
# with open('tweet_json.txt', 'r') as f:
with open('json.txt', 'r') as f:
    gathered_tweet_df = pd.DataFrame(columns=('tweet_id', 'retweet_count', 'favorite_count'))
    tweets = f.readlines()
    for tweet in tweets:
        tweet = json.loads(tweet)
        gathered_tweet_df.loc[len(gathered_tweet_df.index)] = [tweet['id'], tweet['retweet_count'], tweet['favorite_count']]

## Assessing Data
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. You must use **both** visual assessment
programmatic assessement to assess the data.

**Note:** pay attention to the following key points when you access the data.

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.



In [343]:
gathered_tweet_df.shape

(2354, 3)

In [344]:
gathered_tweet_df.sample(4)

tweet_id retweet_count favorite_count
2103  670474236058800128           835           1635
134   866450705531457537         32883         106827
1382  700796979434098688          1085           2669
393   825829644528148480          2848          14025

In [345]:
gathered_tweet_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2354 entries, 0 to 2353
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   tweet_id        2354 non-null   object
 1   retweet_count   2354 non-null   object
 2   favorite_count  2354 non-null   object
dtypes: object(3)
memory usage: 73.6+ KB


In [346]:
gathered_tweet_df.describe()

tweet_id  retweet_count  favorite_count
count                 2354           2354            2354
unique                2354           1724            2007
top     667495797102141441           3652               0
freq                     1              5             179

In [347]:
# Check null values in gathered_tweet_df
gathered_tweet_df.isnull().sum()

tweet_id          0
retweet_count     0
favorite_count    0
dtype: int64

In [348]:
# Check duplicates in gathered_tweet_df
gathered_tweet_df.duplicated().sum()

0

In [349]:
image_pred.sample(4)

tweet_id                                          jpg_url  \
345   672272411274932228  https://pbs.twimg.com/media/CVRkLuJWUAAhhYp.jpg   
835   694183373896572928  https://pbs.twimg.com/media/CaI8Fn0WAAIrFJN.jpg   
475   675149409102012420  https://pbs.twimg.com/media/CV6czeEWEAEdChp.jpg   
1191  739932936087216128  https://pbs.twimg.com/media/CkTFEe-W0AA90m1.jpg   

      img_num       p1   p1_conf  p1_dog                  p2   p2_conf  \
345         2      pug  0.914685    True  Norwegian_elkhound  0.014982   
835         1    teddy  0.441499   False            Pekinese  0.080870   
475         1     chow  0.999876    True     Tibetan_mastiff  0.000059   
1191        1  redbone  0.243904    True              beagle  0.210975   

      p2_dog               p3   p3_conf  p3_dog  
345     True      Siamese_cat  0.009221   False  
835     True         Shih-Tzu  0.072099    True  
475     True  Tibetan_terrier  0.000029    True  
1191    True           vizsla  0.076443    True

In [350]:
image_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [351]:
image_pred.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [352]:
# Check duplicates in image_pred
image_pred.duplicated().sum()

0

In [353]:
# Check null values in tweet_archive
tweet_archive.isnull().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

In [354]:
tweet_archive.sample(4)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2129  670290420111441920                    NaN                  NaN   
244   846042936437604353                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
1023  746521445350707200                    NaN                  NaN   

                      timestamp  \
2129  2015-11-27 17:17:44 +0000   
244   2017-03-26 16:55:29 +0000   
6     2017-07-28 16:27:12 +0000   
1023  2016-06-25 01:52:36 +0000   

                                                                                  source  \
2129  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
244   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
6     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1023  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                        text  \
2129                                                          This is Sandra. She's going skydiving. Nice adidas sandals. Stellar house plant. 11/10 https://t.co/orbkAq9kYF   
244                                                Meet Jarvis. The snow pupsets him. Officially ready for summer. 12/10 would perform a chilly boop https://t.co/0hLkztpiOW   
6     Meet Jax. He enjoys ice cream so much he gets nervous around it. 13/10 help Jax enjoy more things by clicking below\n\nhttps://t.co/Zr4hWfAs1H https://t.co/tVJBRMnhxl   
1023                            RT @dog_rates: This is Shaggy. He knows exactly how to solve the puzzle but can't talk. All he wants to do is help. 10/10 great guy https:/…   

      retweeted_status_id  retweeted_status_user_id  \
2129                  NaN                       NaN   
244                   NaN                       NaN   
6                     NaN                       NaN   
1023         6.678667e+17              4.196984e+09   

     retweeted_status_timestamp  \
2129                        NaN   
244                         NaN   
6                           NaN   
1023  2015-11-21 00:46:50 +0000   

                                                                                                    expanded_urls  \
2129                                              https://twitter.com/dog_rates/status/670290420111441920/photo/1   
244                                               https://twitter.com/dog_rates/status/846042936437604353/photo/1   
6     https://gofundme.com/ydvmve-surgery-for-jax,https://twitter.com/dog_rates/status/890971913173991426/photo/1   
1023                                              https://twitter.com/dog_rates/status/667866724293877760/photo/1   

      rating_numerator  rating_denominator    name doggo floofer pupper puppo  
2129                11                  10  Sandra  None    None   None  None  
244                 12                  10  Jarvis  None    None   None  None  
6                   13                  10     Jax  None    None   None  None  
1023                10                  10  Shaggy  None    None   None  None

In [355]:
tweet_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [356]:
tweet_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [357]:
# Check null values in tweet_archive
tweet_archive.isnull().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

In [358]:
''' Columns such as retweet_status_id have high null values
Those columns and some other ones have high null values and are not useful for our analysis
'''


' Columns such as retweet_status_id have high null values\nThose columns and some other ones have high null values and are not useful for our analysis\n'

In [359]:
# Check duplicates in tweet_archive
tweet_archive.duplicated().sum()

0

In [360]:
tweet_archive.query('doggo == "doggo"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
9     890240255349198849                    NaN                  NaN   
43    884162670584377345                    NaN                  NaN   
99    872967104147763200                    NaN                  NaN   
108   871515927908634625                    NaN                  NaN   
110   871102520638267392                    NaN                  NaN   
...                  ...                    ...                  ...   
1117  732375214819057664                    NaN                  NaN   
1141  727644517743104000                    NaN                  NaN   
1156  724771698126512129                    NaN                  NaN   
1176  719991154352222208                    NaN                  NaN   
1204  716080869887381504                    NaN                  NaN   

                      timestamp  \
9     2017-07-26 15:59:51 +0000   
43    2017-07-09 21:29:42 +0000   
99    2017-06-09 00:02:31 +0000   
108   2017-06-04 23:56:03 +0000   
110   2017-06-03 20:33:19 +0000   
...                         ...   
1117  2016-05-17 01:00:32 +0000   
1141  2016-05-03 23:42:26 +0000   
1156  2016-04-26 01:26:53 +0000   
1176  2016-04-12 20:50:42 +0000   
1204  2016-04-02 01:52:38 +0000   

                                                                                  source  \
9     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
43    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
99    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
108   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
110   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
...                                                                                  ...   
1117  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1141  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1156  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1176  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1204  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                      text  \
9            This is Cassie. She is a college pup. Studying international doggo communication and stick theory. 14/10 so elegant much sophisticate https://t.co/t1bfwz5S2A   
43    Meet Yogi. He doesn't have any important dog meetings today he just enjoys looking his best at all times. 12/10 for dangerously dapper doggo https://t.co/YSI00BzTBZ   
99            Here's a very large dog. He has a date later. Politely asked this water person to check if his breath is bad. 12/10 good to go doggo https://t.co/EMYIdoblMR   
108   This is Napolean. He's a Raggedy East Nicaraguan Zoom Zoom. Runs on one leg. Built for deception. No eyes. Good with kids. 12/10 great doggo https://t.co/PR7B7w1rUw   
110                                                                                                                      Never doubt a doggo 14/10 https://t.co/AbBLh2FZCH   
...                                                                                                                                                                    ...   
1117                             This is Kyle (pronounced 'Mitch'). He strives to be the best doggo he can be. 11/10 would pat on head approvingly https://t.co/aA2GiTGvlE   
1141                                                                                       Here's a doggo struggling to cope with the winds. 13/10 https://t.co/qv3aUwaouT   
1156                                                       

In [361]:
pd.set_option('display.max_colwidth', None)

In [362]:
# Find the tweet with the highest rating numerator
tweet_archive.query('rating_numerator == rating_numerator.max()')['text']

979    This is Atticus. He's quite simply America af. 1776/10 https://t.co/GRXwMxLBkh
Name: text, dtype: object

In [363]:
tweet_archive[['text','name', 'rating_numerator', 'rating_denominator']].sample(15)


text  \
1772                             Another spooky pupper here. Most definitely floating. No legs. Probably knows some dark magic. 10/10 very spooked https://t.co/JK8MByRzgj   
639                                             Oh h*ck look at this spookling right here. Fright level off the charts. 12/10 sufficiently spooked https://t.co/BNy9IIJMb0   
1652                          Meet Banjo. He's a Peppercorn Shoop Da Whoop. Nails look lethal. Skeptical of luminescent orb 11/10 stay woke pupper https://t.co/H7NZFumpKq   
832                           Say hello to Oakley and Charlie. They're convinced that they each have their own stick. Nobody tell them. Both 12/10 https://t.co/J2AJdyxglH   
117   Meet Clifford. He's quite large. Also red. Good w kids. Somehow never steps on them. Massive poops very inconvenient. Still 14/10 would ride https://t.co/apVOyDgOju   
819                            We only rate dogs. Pls stop sending in non-canines like this Arctic Floof Kangaroo. This is very frustrating. 11/10 https://t.co/qlUDuPoE3d   
217   This is Leo. He's a personal triathlon coach. Currently overseeing this athlete's push-pups. H*ckin brutal. 13/10 would do all he asks of me https://t.co/FXZQtBcnTO   
778                                                                              RT @dog_rates: Like father (doggo), like son (pupper). Both 12/10 https://t.co/pG2inLaOda   
992                            That is Quizno. This is his beach. He does not tolerate human shenanigans on his beach. 10/10 reclaim ur land doggo https://t.co/vdr7DaRSa7   
1511                                                           I present to you... Dog Jesus. 13/10 (he could be sitting on a rock but I doubt it) https://t.co/fR1P3g5I6k   
240            This is Barney. He's an elder doggo. Hitches a ride when he gets tired. Waves goodbye before he leaves. 13/10 please come back soon https://t.co/cFAasDXauK   
793                                                                           This is Chelsea. She forgot how to dog. 11/10 get it together pupper https://t.co/nBJ5RE4yHb   
1213                                        This is Nico. His selfie game is strong af. Excellent use of a sneaky tongue slip. 10/10 star material https://t.co/1OBdJkMOFx   
1559                                                                           This is Frönq. He got caught stealing a waffle. Damn it Frönq. 9/10 https://t.co/7ycWCUrjmZ   
1033                                           This is Jeffrey. He wasn't prepared to execute such advanced barkour. Still 11/10 would totally pet https://t.co/MuuwkkLrHh   

          name  rating_numerator  rating_denominator  
1772      None                10                  10  
639       None                12                  10  
1652     Banjo                11                  10  
832     Oakley                12                  10  
117   Clifford                14                  10  
819       very                11                  10  
217        Leo                13                  10  
778       None                12                  10  
992        his                10                  10  
1511      None                13                  10  
240     Barney                13                  10  
793    Chelsea                11                  10  
1213      Nico                10                  10  
1559     Frönq                 9                  10  
1033   Jeffrey                11                  10

In [364]:
tweet_archive[['text', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp']].sample(15)

text  \
1844                                                     This dog is being demoted to a 9/10 for not wearing a helmet while riding. Gotta stay safe out there. Thank you   
1363                           This is Chip. He's an Upper West Nile Pantaloon. Extremely deadly. Will rip your throat out. 6/10 might still pet https://t.co/LUFnwzznaV   
1677                                                              This is Tess. Her main passions are shelves and baking too many cookies. 11/10 https://t.co/IriJlVZ6m4   
2073                                                                       Yea I can't handle this job anymore your dogs are too adorable. 12/10 https://t.co/N9W5L7BLTm   
545                                                                                       This is Duke. He is not a fan of the pupporazzi. 12/10 https://t.co/SgpBVYIL18   
2198                           This is a wild Toblerone from Papua New Guinea. Mouth always open. Addicted to hay. Acts blind. 7/10 handsome dog https://t.co/IGmVbz07tZ   
2306                    These are Peruvian Feldspars. Their names are Cupit and Prencer. Both resemble Rand Paul. Sick outfits 10/10 &amp; 10/10 https://t.co/ZnEMHBsAs1   
531                     Here we have Burke (pupper) and Dexter (doggo). Pupper wants to be exactly like doggo. Both 12/10 would pet at same time https://t.co/ANBpEYHaho   
1213                                      This is Nico. His selfie game is strong af. Excellent use of a sneaky tongue slip. 10/10 star material https://t.co/1OBdJkMOFx   
962                         Meet Milo. He hauled ass until he ran out of treadmill and then passed out from exhaustion. 11/10 sleep tight pupper https://t.co/xe1aGZNkcC   
776                                                 This is Sampson. He's about to get hit with a vicious draw 2. Has no idea. 11/10 poor pupper https://t.co/FYT9QBEnKG   
12                                    Here's a puppo that seems to be on the fence about something haha no but seriously someone help her. 13/10 https://t.co/BxvuXk0UCm   
405                                                              This is Cash. He's officially given pup on today. 12/10 frighteningly relatable https://t.co/m0hrATIEyw   
1220                                                                         Here are two lil cuddly puppers. Both 12/10 would snug like so much https://t.co/zO4eb7C4tG   
193   Guys, we only rate dogs. This is quite clearly a bulbasaur. Please only send dogs. Thank you... 12/10 human used pet, it's super effective https://t.co/Xc7uj1C64x   

      retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  
1844                  NaN                       NaN                        NaN  
1363                  NaN                       NaN                        NaN  
1677                  NaN                       NaN                        NaN  
2073                  NaN                       NaN                        NaN  
545                   NaN                       NaN                        NaN  
2198                  NaN                       NaN                        NaN  
2306                  NaN                       NaN                        NaN  
531                   NaN                       NaN                        NaN  
1213                  NaN                       NaN                        NaN  
962                   NaN                       NaN                        NaN  
776                   NaN                       NaN                        NaN  
12                    NaN                       NaN                        NaN  
405                   NaN                       NaN                        NaN  
1220                  NaN                       NaN                        NaN  
193                   NaN                       NaN                        NaN

In [365]:
tweet_archive['name'].value_counts()

None         745
a             55
Charlie       12
Oliver        11
Lucy          11
            ... 
Brandi         1
Chubbs         1
Tedrick        1
Torque         1
Enchilada      1
Name: name, Length: 957, dtype: int64

### Quality issues
``tweet_archive table``

1. Some of the tweets are retweets and some are not even about dogs and still have ratings

2. Some of the columns like in_reply_to_status_id, in_reply_to_user_id have no real use case and are motly null

3. Some of the dog names are incorrect and some of them having the value None

4. Incorrect ratings for some of the dogs

5. Incorrect data type for some of the columns like timestamp

``image_pred table``

6. Tweets with false p1_dog value tend not to be dog

7. Image number column doesn't seem to convey any actual value for analysis

8. Wrong data type for p1, p2, p3

### Tidiness issues
1. The dog stages should have been a single column instead of being split into three

2. Too many datasets. They can be merged for managability

## Cleaning Data
In this section, clean **all** of the issues you documented while assessing. 

**Note:** Make a copy of the original data before cleaning. Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). The result should be a high-quality and tidy master pandas DataFrame (or DataFrames, if appropriate).

In [366]:
# Make copies of original pieces of data
tweet_archive_copy = tweet_archive.copy()
image_pred_copy = image_pred.copy()
gathered_tweet_df_copy = gathered_tweet_df.copy()

In [367]:
dogitionary = ['doggo', 'floofer', 'pupper', 'puppo']

### Issue #1:
* Some of the tweets are retweets and may not be about dogs

#### Define:
- Tweets having non-null values in retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp should be dropped
- It is noticed from the describe function above we have a total of 181 non-null values in these columns

#### Code

In [368]:
tweet_archive_copy.shape

(2356, 17)

In [369]:
# Drop rows having non-null values in retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp columns of tweet_archive_copy
tweet_archive_copy = tweet_archive_copy.loc[tweet_archive_copy['retweeted_status_id'].isnull()]

In [370]:
# Drop the retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp columns of tweet_archive_copy
tweet_archive_copy = tweet_archive_copy.drop(['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], axis=1)

#### Test

In [371]:
tweet_archive_copy.shape


(2175, 14)

In [372]:
tweet_archive_copy.sample(4)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1051  742534281772302336                    NaN                  NaN   
129   867421006826221569                    NaN                  NaN   
1677  682059653698686977                    NaN                  NaN   
2110  670444955656130560                    NaN                  NaN   

                      timestamp  \
1051  2016-06-14 01:49:03 +0000   
129   2017-05-24 16:44:18 +0000   
1677  2015-12-30 04:44:28 +0000   
2110  2015-11-28 03:31:48 +0000   

                                                                                  source  \
1051                     <a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>   
129   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1677  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2110  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                               text  \
1051                                              For anyone who's wondering, this is what happens after a doggo catches it's tail... 11/10 https://t.co/G4fNhzelDv   
129   This is Shikha. She just watched you drop a skittle on the ground and still eat it. Could not be less impressed. 12/10 superior puppo https://t.co/XZlZKd73go   
1677                                                         This is Tess. Her main passions are shelves and baking too many cookies. 11/10 https://t.co/IriJlVZ6m4   
2110                                                                       This is Paull. He just stubbed his toe. 10/10 deep breaths Paull https://t.co/J5Mqn8VeYq   

                                                                                                                        expanded_urls  \
1051                                                                                                    https://vine.co/v/iLTZmtE1FTB   
129                                                                   https://twitter.com/dog_rates/status/867421006826221569/photo/1   
1677  https://twitter.com/dog_rates/status/682059653698686977/photo/1,https://twitter.com/dog_rates/status/682059653698686977/photo/1   
2110                                                                  https://twitter.com/dog_rates/status/670444955656130560/photo/1   

      rating_numerator  rating_denominator    name  doggo floofer pupper  \
1051                11                  10    None  doggo    None   None   
129                 12                  10  Shikha   None    None   None   
1677                11                  10    Tess   None    None   None   
2110                10                  10   Paull   None    None   None   

      puppo  
1051   None  
129   puppo  
1677   None  
2110   None

### Issue #2:
- Invalid columns with almost all NaN values

#### Define:
- in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp,source
- Drop above columns with the drop function

#### Code

In [373]:
useless_columns = ['in_reply_to_status_id', 'in_reply_to_user_id','source']

In [374]:
tweet_archive_copy.drop(useless_columns, axis=1, inplace=True)

#### Test

In [375]:
tweet_archive_copy.sample(4)

tweet_id                  timestamp  \
1635  684222868335505415  2016-01-05 04:00:18 +0000   
1949  673689733134946305  2015-12-07 02:25:23 +0000   
1259  710272297844797440  2016-03-17 01:11:26 +0000   
839   767191397493538821  2016-08-21 02:47:37 +0000   

                                                                                                                                              text  \
1635  Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110 https://t.co/1zfnTJLt55   
1949                                                      When you're having a blast and remember tomorrow's Monday. 11/10 https://t.co/YPsJasNVGe   
1259    We 👏🏻 only 👏🏻 rate 👏🏻 dogs. Pls stop sending in non-canines like this Dutch Panda Worm. This is infuriating. 11/10 https://t.co/odfLzBonG2   
839        I don't know any of the backstory behind this picture but for some reason I'm crying. 13/10 for owner and doggo https://t.co/QOKZdus9TT   

                                                                                                                        expanded_urls  \
1635                                                                  https://twitter.com/dog_rates/status/684222868335505415/photo/1   
1949  https://twitter.com/dog_rates/status/673689733134946305/photo/1,https://twitter.com/dog_rates/status/673689733134946305/photo/1   
1259                                                                  https://twitter.com/dog_rates/status/710272297844797440/photo/1   
839                                                                   https://twitter.com/dog_rates/status/767191397493538821/photo/1   

      rating_numerator  rating_denominator         name  doggo floofer pupper  \
1635               121                 110         None   None    None   None   
1949                11                  10         None   None    None   None   
1259                11                  10  infuriating   None    None   None   
839                 13                  10         None  doggo    None   None   

     puppo  
1635  None  
1949  None  
1259  None  
839   None

### Issue #3:
- Incorrect names
- None values for some of the names

#### Define:
- Find the names that are not correct by using value count
- Replace incorrect names and None values with NaN

#### Code

In [376]:
# # First, remove all tweets that don't contain any of the dog words
# for word in dogitionary:
#     tweet_archive_copy = tweet_archive_copy[tweet_archive_copy['text'].str.contains(word)]


In [377]:
# Create a csv file containg names of dogs and view them visually
counts = tweet_archive_copy['name'].value_counts()
counts.to_csv('name.csv', index=True)

In [378]:
# Get all the invalid names and remove them from the dataframe
# We notice invalid names starts with lowercase letters.

# Create a list of invalid names
invalid_names = ['None']
for name in tweet_archive_copy.name:
    if name[0].islower():
        invalid_names.append(name)

In [379]:
# Get unique invalid names
invalid_names = list(set(invalid_names))

In [380]:
tweet_archive_copy.shape

(2175, 11)

In [381]:
# Remove invalid names from the dataframe
tweet_archive_copy = tweet_archive_copy[~tweet_archive_copy['name'].isin(invalid_names)]
tweet_archive_copy.name.value_counts()

Charlie      11
Lucy         11
Oliver       10
Cooper       10
Penny         9
             ..
Durg          1
Grey          1
Tripp         1
Glenn         1
Enchilada     1
Name: name, Length: 930, dtype: int64

In [382]:
# View the dataframe
tweet_archive_copy.shape

(1391, 11)

#### Test

In [383]:
# verify that the dataframe is now clean of invalid names
tweet_archive_copy[['text','name']].sample(10)

text  \
721                      This is Dave. He's currently in a predicament. Doesn't seem to mind tho. 12/10 someone assist Dave https://t.co/nfprKAXqwu   
715   This is Riley. His owner put a donut pillow around him and he loves it so much he won't let anyone take it off. 13/10 https://t.co/8TCQcsZCZ8   
1796                                            This is Juckson. He's totally on his way to a nascar race. 5/10 for Juckson https://t.co/IoLRvF0Kak   
1701                                                                                    This is Alice. She's an idiot. 4/10 https://t.co/VQXdwJfkyS   
65                               This is Bailey. He thinks you should measure ear length for signs of growth instead. 12/10 https://t.co/IxM9IMKQq8   
898          This is Lilli Bee &amp; Honey Bear. Unfortunately, they were both born with no eyes. So heckin sad. Both 11/10 https://t.co/4UrfOZhztW   
625                                                 This is Brody. He's trying to make the same face as the football. 12/10 https://t.co/2H4MfOGlpQ   
1704            This is Cheesy. It's her birthday. She's patiently waiting to eat her party muffin. 9/10 happy birthday pup https://t.co/cH2H7mch2H   
2118                                       This is Damon. The newest presidential candidate for 2016. 10/10 he gets my vote https://t.co/Z5nqlfjYJi   
2100        Meet Danny. He's too good to look at the road when he's driving. Absolute menace. 6/10 completely irresponsible https://t.co/I1lMUy1FqH   

         name  
721      Dave  
715     Riley  
1796  Juckson  
1701    Alice  
65     Bailey  
898     Lilli  
625     Brody  
1704   Cheesy  
2118    Damon  
2100    Danny

### Issue #4:
Incorrect Ratings for some of the dogs

#### Define
- We were told the denominator is always 10. By viewing the describe function above we can confirm the denominator has
- numbers greater than 10
- We will find all numbers greater than 10 in the denominator column and replace them with 10.
- We will also find uncommon numerators and replace them with proper values

#### Code

In [384]:
# Reset pandas display options
pd.reset_option('display.max_colwidth')

In [385]:
tweet_archive_copy[['text','name', 'rating_numerator', 'rating_denominator']].sample(10)

text     name  \
810   Meet Fizz. She thinks love is a social constru...     Fizz   
1442  This is Reptar. He specifically asked for his ...   Reptar   
1581  Say hello to Samson. He's a Firecracker Häagen...   Samson   
1742  Meet Hurley. He's the curly one. He hugs every...   Hurley   
827   This is DonDon. He's way up but doesn't feel b...   DonDon   
932   This is Charlie. He pouts until he gets to go ...  Charlie   
16    This is Jim. He found a fren. Taught him how t...      Jim   
950   This is Brody. He's a lifeguard. Always prepar...    Brody   
2121  This is Peanut. He was the World Table Tennis ...   Peanut   
239   This is Tuck. As you can see, he's rather h*ck...     Tuck   

      rating_numerator  rating_denominator  
810                 11                  10  
1442                10                  10  
1581                11                  10  
1742                11                  10  
827                 12                  10  
932                 12                  10  
16                  12                  10  
950                 12                  10  
2121                10                  10  
239                 13                  10

In [386]:
# Find the distribution of denominator values
tweet_archive_copy['rating_denominator'].value_counts()

10    1388
50       1
11       1
7        1
Name: rating_denominator, dtype: int64

In [387]:
tweet_archive_copy.query('rating_denominator < 10')

tweet_id                  timestamp  \
516  810984652412424192  2016-12-19 23:06:23 +0000   

                                                  text  \
516  Meet Sam. She smiles 24/7 &amp; secretly aspir...   

                                         expanded_urls  rating_numerator  \
516  https://www.gofundme.com/sams-smile,https://tw...                24   

     rating_denominator name doggo floofer pupper puppo  
516                   7  Sam  None    None   None  None

In [388]:
tweet_archive_copy.query('rating_denominator > 10')

tweet_id                  timestamp  \
1202  716439118184652801  2016-04-03 01:36:11 +0000   
1662  682962037429899265  2016-01-01 16:30:13 +0000   

                                                   text  \
1202  This is Bluebert. He just saw that both #Final...   
1662  This is Darrel. He just robbed a 7/11 and is i...   

                                          expanded_urls  rating_numerator  \
1202  https://twitter.com/dog_rates/status/716439118...                50   
1662  https://twitter.com/dog_rates/status/682962037...                 7   

      rating_denominator      name doggo floofer pupper puppo  
1202                  50  Bluebert  None    None   None  None  
1662                  11    Darrel  None    None   None  None

In [389]:
# View the distribution of rating numerator
tweet_archive_copy.rating_numerator.describe()

count    1391.000000
mean       12.091301
std        47.413241
min         2.000000
25%        10.000000
50%        11.000000
75%        12.000000
max      1776.000000
Name: rating_numerator, dtype: float64

In [390]:
# Find all ratings numerator greater than the 75th percentile
greater_than_75 = tweet_archive_copy['rating_numerator'][tweet_archive_copy['rating_numerator'] > tweet_archive_copy['rating_numerator'].quantile(0.75)]
print(greater_than_75.value_counts())

13      183
14       17
1776      1
75        1
50        1
27        1
24        1
Name: rating_numerator, dtype: int64


In [391]:
pd.set_option('display.max_colwidth', None)

In [392]:
# Find the tweet with the highest rating numerator
tweet_archive_copy.query('rating_numerator == rating_numerator.max()')

tweet_id                  timestamp  \
979  749981277374128128  2016-07-04 15:00:45 +0000   

                                                                               text  \
979  This is Atticus. He's quite simply America af. 1776/10 https://t.co/GRXwMxLBkh   

                                                       expanded_urls  \
979  https://twitter.com/dog_rates/status/749981277374128128/photo/1   

     rating_numerator  rating_denominator     name doggo floofer pupper puppo  
979              1776                  10  Atticus  None    None   None  None

In [393]:
# The above tweet has a rating numerator greater than the 75th percentile and seems to somehow correlate to the American Independence Day. 
# We can drop this tweet from the dataframe since it is an outlier and we don't want to include it in our analysis.
tweet_archive_copy = tweet_archive_copy[tweet_archive_copy['rating_numerator'] != 1776]

In [394]:
tweet_archive_copy.query('rating_numerator == "75"')

tweet_id                  timestamp  \
695  786709082849828864  2016-10-13 23:23:56 +0000   

                                                                                                                                 text  \
695  This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS   

                                                       expanded_urls  \
695  https://twitter.com/dog_rates/status/786709082849828864/photo/1   

     rating_numerator  rating_denominator   name doggo floofer pupper puppo  
695                75                  10  Logan  None    None   None  None

In [395]:
# On close analysis we find that the extracted ratings numerator is wrong. We need to correct the rating.
# We can correct the ratings by using the approximated figure in the tweet.

# Replace the rating numerator with the approximated figure
tweet_archive_copy['rating_numerator'] = tweet_archive_copy['rating_numerator'].apply(lambda x: 10 if x==75 else x)

In [396]:
tweet_archive_copy.query('rating_numerator == "50"')

tweet_id                  timestamp  \
1202  716439118184652801  2016-04-03 01:36:11 +0000   

                                                                                                                       text  \
1202  This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq   

                                                        expanded_urls  \
1202  https://twitter.com/dog_rates/status/716439118184652801/photo/1   

      rating_numerator  rating_denominator      name doggo floofer pupper  \
1202                50                  50  Bluebert  None    None   None   

     puppo  
1202  None

In [397]:
# Replace the numerator and denominator ratings having a value of 50 with the exact figure
tweet_archive_copy['rating_numerator'] = tweet_archive_copy['rating_numerator'].apply(lambda x: 11 if x==50 else x)
tweet_archive_copy['rating_denominator'] = tweet_archive_copy['rating_denominator'].apply(lambda x: 10 if x==50 else x)


In [398]:
tweet_archive_copy.query('rating_numerator == "27"')

tweet_id                  timestamp  \
763  778027034220126208  2016-09-20 00:24:34 +0000   

                                                                                                                                                                  text  \
763  This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back https://t.co/QFaUiIHxHq   

                                                       expanded_urls  \
763  https://twitter.com/dog_rates/status/778027034220126208/photo/1   

     rating_numerator  rating_denominator    name doggo floofer  pupper puppo  
763                27                  10  Sophie  None    None  pupper  None

In [399]:
# Using the above manual process we can correct the ratings numerator to the approximated figure.
tweet_archive_copy['rating_numerator'] = tweet_archive_copy['rating_numerator'].apply(lambda x: 11 if x==27 else x)

In [400]:
tweet_archive_copy.query('rating_numerator == "24"')

tweet_id                  timestamp  \
516  810984652412424192  2016-12-19 23:06:23 +0000   

                                                                                                                                                                          text  \
516  Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx   

                                                                                           expanded_urls  \
516  https://www.gofundme.com/sams-smile,https://twitter.com/dog_rates/status/810984652412424192/photo/1   

     rating_numerator  rating_denominator name doggo floofer pupper puppo  
516                24                   7  Sam  None    None   None  None

In [401]:
# Drop row with rating numerator of 24
tweet_archive_copy = tweet_archive_copy[tweet_archive_copy['rating_numerator'] != 24]

#### Test

In [402]:
# Print the highest and lowest rating denominator and numerator
tweet_archive_copy.groupby('rating_denominator').rating_numerator.max().sort_values(ascending=False)



rating_denominator
10    14
11     7
Name: rating_numerator, dtype: int64

In [403]:
tweet_archive_copy.groupby('rating_numerator').rating_denominator.max().sort_values(ascending=False)


rating_numerator
7     11
14    10
13    10
12    10
11    10
10    10
9     10
8     10
6     10
5     10
4     10
3     10
2     10
Name: rating_denominator, dtype: int64

### Issue 5:
- Some columns have wrong datatypes

#### Define
- Change datatypes for columns such as timestamp using pandas datetime function

#### Code

In [404]:
# Confirm datatypes of columns
tweet_archive_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1389 entries, 0 to 2325
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            1389 non-null   int64 
 1   timestamp           1389 non-null   object
 2   text                1389 non-null   object
 3   expanded_urls       1389 non-null   object
 4   rating_numerator    1389 non-null   int64 
 5   rating_denominator  1389 non-null   int64 
 6   name                1389 non-null   object
 7   doggo               1389 non-null   object
 8   floofer             1389 non-null   object
 9   pupper              1389 non-null   object
 10  puppo               1389 non-null   object
dtypes: int64(3), object(8)
memory usage: 130.2+ KB


In [405]:
# Change the datatype for timestamp column to datetime
tweet_archive_copy['timestamp'] = pd.to_datetime(tweet_archive_copy['timestamp'])

In [406]:
# Find distribution of tweets by year of creation in tweet_archive_copy
tweet_archive_copy['timestamp'].dt.year.value_counts()

2016    726
2015    379
2017    284
Name: timestamp, dtype: int64

#### Test

In [407]:
tweet_archive_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1389 entries, 0 to 2325
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            1389 non-null   int64              
 1   timestamp           1389 non-null   datetime64[ns, UTC]
 2   text                1389 non-null   object             
 3   expanded_urls       1389 non-null   object             
 4   rating_numerator    1389 non-null   int64              
 5   rating_denominator  1389 non-null   int64              
 6   name                1389 non-null   object             
 7   doggo               1389 non-null   object             
 8   floofer             1389 non-null   object             
 9   pupper              1389 non-null   object             
 10  puppo               1389 non-null   object             
dtypes: datetime64[ns, UTC](1), int64(3), object(7)
memory usage: 130.2+ KB


### Issue 6:
- Tweets with false p1_dog value tend not to be dogs

#### Define
- Delete rows containing False as a value in p1_dog column

#### Code

In [408]:
# Delete rows containing False in the p1_dog column in the image_pred_copy dataframe
image_pred_copy = image_pred_copy[image_pred_copy['p1_dog'] == True]

#### Test

In [ ]:
image_pred_copy.groupby('p1_dog').count()

In [ ]:
image_pred_copy.shape

### Issue 7:
- img_num column does'nt seem to convey any actual value for analys

#### Define
- img_num column should be dropped using the drop method

#### Code

In [ ]:
# Drop image_num column from image_pred
image_pred_copy = image_pred_copy.drop(['img_num'], axis=1)

<?@>

#### Test

In [ ]:
image_pred_copy.sample(4)

### Issue 8:
- Wrong data type for p1, p2, p3

#### Define
- Convert p1, p2, p3 data type from strings to categorical type

#### Code

In [ ]:
# Convert p1, p2, p3 datatype to categorical

convert_dict = {'p1': 'category', 'p2': 'category', 'p3': 'category'}
image_pred_copy = image_pred_copy.astype(convert_dict)

#### Test

In [ ]:
image_pred_copy.info()

### Tidiness Issue

### Issue 1
- The three dog stages is not necessary. There should all be under a single column

#### Define
- Collapse the three dog stages into a single column
- Use the pandas melt function

In [ ]:
# Use pandas melt function to convert the dataframe to long format
tweet_archive_copy = pd.melt(tweet_archive_copy, id_vars=['tweet_id','timestamp','text','expanded_urls','rating_numerator','rating_denominator','name'], value_vars=['doggo','floofer', 'pupper','puppo'], var_name='dog_type')

In [ ]:
tweet_archive_copy.drop('value', axis=1, inplace=True)

#### Test

In [ ]:
tweet_archive_copy.sample(4)

In [ ]:
tweet_archive_copy.shape

### Issue 2:
- Three datasets instead of one

#### Define
- The three datasets can be merged together using pandas merge function

#### Code

In [ ]:
from functools import reduce

#define list of DataFrames
dfs = [tweet_archive_copy, image_pred_copy, gathered_tweet_df]

#merge all DataFrames into one
final_df = reduce(lambda  left,right: pd.merge(left,right,on=['tweet_id'],
                                            how='outer'), dfs)

In [ ]:
final_df.shape

In [ ]:
# Drop rows with NaN values in final_df
final_df = final_df.dropna()

In [ ]:
final_df.sample(4)

In [ ]:
final_df.shape

In [ ]:
final_df.info()

### Extra Issues in the final df
- integer values for tweet id
- ratings should int64
- dog_type should be a category
- retweets and favorite counts should be int64

#### Define
- Convert datatypes of the affected columns into the appropriate type

#### Code

In [ ]:
# Create a dictionary for the datatypes of the columns
final_convert_dict = {'tweet_id': 'int64', 'rating_numerator': 'int64', 
                        'rating_denominator': 'int64', 'dog_type': 'category', 
                        'retweet_count':'int64', 'favorite_count':'int64'}
final_df = final_df.astype(final_convert_dict)

#### Test

In [ ]:
final_df.info()

In [ ]:
final_df.sample(2)

## Storing Data
Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

In [ ]:
# Save final_df to csv
final_df.to_csv('twitter_archive_master.csv', index=False)

## Analyzing and Visualizing Data
In this section, analyze and visualize your wrangled data. You must produce at least **three (3) insights and one (1) visualization.**

#### Questions

1 How active was the WeRateDogs page in the different years
1 Are some dog types more pouplar than others?
3 What are the most common breed of dogs?
2 What breed of dogs had the highest likes and and retweets
3 What are the dogs with the highest numerator rating
4 What breed of dogs are the most common
5 The neural networks had the best performance or confidence on what breed of dogs

##### 1

In [ ]:
# View numbers of tweets by year
final_df['timestamp'].dt.year.value_counts()

In [ ]:
# Plot a visualization of the number of tweets by year
final_df['timestamp'].dt.year.value_counts().plot(kind='bar')

#### 2

In [ ]:
# Plot the dog type distribution in the final_df
final_df['dog_type'].value_counts().plot(kind='bar')

#### 3

In [ ]:
# Find the top 50 most common dog breeds
top_50_breeds = final_df['p1'].value_counts().head(50)

In [ ]:
# Plot a bar chart of the top 50 most common dog breeds
plt.figure(figsize = (20, 10))
  
ax = sns.barplot(x=top_50_breeds.index, y=top_50_breeds, palette='husl')
ax.set_xticklabels(ax.get_xticklabels(), rotation=90);

#### 4

In [ ]:
# Group by dog_type and count the number of favorites and retweets
like_df = final_df.groupby('dog_type').agg({'favorite_count':'sum', 'retweet_count':'sum'})

In [ ]:
sns.scatterplot(x='retweet_count', y='favorite_count', data=like_df)


In [ ]:
# sum values of retweet_count and favorite_count in like_df
like_df['total_likes'] = like_df['retweet_count'] + like_df['favorite_count']

In [ ]:
# Plot dog type against total likes
like_df.plot(kind='bar', x='dog_type', y='total_likes')

### Insights:
1. 

2.

3.

### Visualization